nexus/src/external_api/http_entrypoints.rs

```diff
@@ -5548,6 +5550,56 @@ async fn silo_metric(
     apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
 }
 
+/// List available timeseries schema.
+#[endpoint {
+    method = GET,
+    path = "/v1/timeseries/schema",
+    tags = ["metrics"],
+}]
+async fn timeseries_schema_list(
+    rqctx: RequestContext<Arc<ServerContext>>,
+    pag_params: Query<oximeter_db::TimeseriesSchemaPaginationParams>,
+) -> Result<HttpResponseOk<ResultsPage<oximeter_db::TimeseriesSchema>>, HttpError>
+{
+    let apictx = rqctx.context();
+    let handler = async {
+        let nexus = &apictx.nexus;
+        let opctx = crate::context::op_context_for_external_api(&rqctx).await?;
+        let pagination = pag_params.into_inner();
+        let limit = rqctx.page_limit(&pagination)?;
+        nexus
+            .timeseries_schema_list(&opctx, &pagination, limit)
+            .await
+            .map(HttpResponseOk)
+            .map_err(HttpError::from)
+    };
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+/// Run a timeseries query, written OxQL.
+#[endpoint {
+    method = POST,
+    path = "/v1/timeseries/query",
+    tags = ["metrics"],
+}]
+async fn timeseries_query(
```

Comment:
Yeah, I definitely avoided this. I think I'm just unsure what to paginate! This returns a list of tables, each containing a list of timeseries, each with a number of data points. I think we probably want to paginate the innermost items, the points themselves. But that seems to imply we either don't exactly respect the pagination limit, or we cut off timeseries partway through? Neither felt satisfying to me. Let me know if you have ideas, and also cc @ahl

Reply:
I talked with @ahl for a bit about this. The story turns out to be quite complicated for this type of object we're returning the API.

Page sizes are _supposed_ to refer to the top-level object, which would be the `Table` here. That's not very useful, since those are containers. We would need to flatten the API for that to work well, returning something closer to a list of _points_. But that would require a bunch of duplication -- we would want to attach the information in a `Table` and `Timeseries` for each of those points, so every point in a timeseries would have the same values for those.

But I think this in and of itself isn't sufficient. Queries can be arbitrarily complicated. I think we'd need to run the query; generate all the data; and _then_ take the first chunk based on the page size. This is really tricky! Queries can look benign and actually fetch a small amount of data from the DB, but then blow it up massively. For example, suppose you fetch 2 samples from a timeseries sampled at 1s, and then align it to a 1ns window. Well, now you have 2 billion points!

If the main goal of pagination is to avoid DOS attacks, then we can't run the query but only return the first N. We need to check in many places how many points we're _going_ to generate, and make sure it's reasonable. This also interacts poorly with fetching the _next_ page of results. We don't have a cursor, so we're stuck with limit-offset pagination, which means we might generate a whole bunch of pages only to throw away all but the last of them.

At the end of the day, I think this is going to be a tricky topic that will take some time to get right. For now, we have two limits in place: You can't fetch more than 100k rows at a time from the DB; and you can't make a query _arbitrarily_ complicated, because they're limited to 4096 characters. I am going to change those limits up a bit, and instead keep track of the total number of samples we've retrieved during a query, and limit that to something reasonable (1M or 10M, maybe). I will then _also_ add something that prevents the issue I mentioned above, where alignment can balloon the data enormously. One way to do that is to prevent upsampling by more than some fixed amount, (say 10x). All of this should provide some crude but reasonable guardrails to prevent huge allocations.

I'm also going to make an issue, and possibly an RFD, to track all this work, since it's likely to be ongoing and complicated.

Reply:
Ok, I've added these two new query-complexity limits in b9dbd8f53.

They check that we don't upsample the data by more than a factor of 10, and they limit the total number of results we're returning from ClickHouse to 1M. So we should never get more than 10M data points back from a query, no matter what. The failure mode is that we just fail the query entirely, which isn't great, but it's much more robust than the previous versions.

I've also added a nice improvement on that row limit over the last commit. We were making the measurements query twice: once to get the number of rows it would return, and then again to actually return them. That is inefficient, and a TOCTOU. More rows could be added after the check, though I didn't think it was too likely to be that bad.

 But this commit instead puts a limit on the number of results, at 1 more than our current remaining allotment. If we get that many, then we know there are too many results, even if we don't know how many more there are. So we can fail the query after running the SQL query just once, and while returning at most one more row than we want to.